### ISSUE TITLE
Auto-complete and auto-import multi-segment path

### ISSUE THREAD

**Issue by pcpthm** [rust-analyzer/rust-analyzer#1062]
One of the most convenient features I'm missing for the IDE experience is ability to import any name as I type.

```rust
fn main() {
    stdi<|>
}
```
to
```rust
use std::io::stdin;

fn main() {
    stdin()<|>
}
```

Requirements:
- Names which are in the scope should be prioritized (shown above in the completion list) over names not in the scope.
- If multiple paths corresponds to an item, completion result shouldn't be duplicated.

**Comment by matklad** [comment]
This is not too close to being implemented: we need to implement macros 90% correctly to have reasonable amount of false-positives. But this is definitely on the long-term roadmap!

**Comment by eulerdisk** [comment]
I was already working on this some weeks ago, I had a working prototype with all the boilerplate and dummy importable items.
[Since then some refactoring were made, so all that code now is probably broken.]

The main missing piece of logic was how to get importable items. My idea was to start with items from external crate doing something like this:
- get the current crate id (for the current file), and its dependencies (@matklad that should already include even ```std``` right?) 
- create a query to get a list of "importable items" for a give crate id.
(this list ideally should be indexed with something like ```fst```). @matklad should this use name resolution?

**Comment by matklad** [comment]
>that should already include even std right?

yes

>create a query to get a list of "importable items" for a give crate id.
(this list ideally should be indexed with something like fst). 

Yeah, I think this is how this should work. Specifically, we should collect all `pub` items, but probably without paths. When a user actually completes a `pub` item, we need to figure out the path under which the def is known

**Comment by flodiebold** [comment]
I think implementing this with a hardcoded set of names from std would probably be a good way to get started, and immediately useful, so the item collection could be done as a separate second step...

**Comment by eulerdisk** [comment]
Then my old branch was already good for that, I need only to rebase.
I will make a PR next week probably.

**Comment by Geobert** [comment]
This is great! We should also support the opposite: I'm doing some Go with vscode for the first time and when I save the file, it automatically removes unused imports. Neat ^^

**Comment by bjorn3** [comment]
@Geobert What if upstream already had those unused imports, and you dont want to have unrelated changes in a PR you are making. Also a proc macro or include! may have used those imports without ra knowing it.

**Comment by lnicola** [comment]
Go errors on unused imports, IIRC. Rust only gives a warning.

**Comment by Geobert** [comment]
> 
> 
> @Geobert What if upstream already had those unused imports, and you dont want to have unrelated changes in a PR you are making. Also a proc macro or include! may have used those imports without ra knowing it.

If `ra` becomes the way to have IDE support, it won't happen as the uses will be keep clean ^^ There might be some transition period where this fix could be optional :)

**Comment by bjorn3** [comment]
> If ra becomes the way to have IDE support, it won't happen as the uses will be keep clean ^^

What about merging two branches each removing part of the uses. And what about a plain old text editor. I think using `cargo fix` the same way as `cargo fmt` is used would be more useful.

**Comment by Geobert** [comment]
> What about merging two branches each removing part of the uses. And what about a plain old text editor. I think using `cargo fix` the same way as `cargo fmt` is used would be more useful.

`cargo fix` can do that?? I didn't know!

**Comment by bjorn3** [comment]
I dont think there is actually an option to fix warnings (I assumed there is), but `#![deny(warnings)]` should make it fix all errors.

**Comment by i-schuetz** [comment]
Is it documented anywhere how this works exactly? If works as expected with imports from my project, but not for example for `std::thread`.

**Comment by bjorn3** [comment]
Does adding `use std::thread;` and hovering over `thread` show the documentation for `std::thread`?

**Comment by i-schuetz** [comment]
@bjorn3 yeah

**Comment by SomeoneToIgnore** [comment]
> Is it documented anywhere how this works exactly? 

I've documented the feature in the next PR that touches it: https://github.com/rust-analyzer/rust-analyzer/pull/6706/commits/c34ac0d62820c82543bb811cc0df48831bb0e906

Currently we don't propose the modules in such completions at all, to avoid showing too many completion options (and computing too much data too, since it might slow things down significantly in some cases).
